lib/src/Uncertainty/Algorithm/Sensitivity/HSICEstimatorImplementation.cxx

```diff
@@ -0,0 +1,434 @@
+  inputDimension_(X.getDimension()),
+  HSIC_XY_ (Point()),
+  HSIC_XX_ (Point()),
+  HSIC_YY_ (Point()),
+  R2HSICIndices_ (Point()),
+  permutationSize_(100)
+{
+  if(covarianceList_.getSize() != (inputSample_.getDimension() + outputSample_.getDimension())) throw InvalidDimensionException(HERE) << "The number of covariance momdels is the dimension of the input +1";
+  if(outputSample_.getDimension() != 1) throw InvalidDimensionException(HERE) << "The dimension of the output is 1.";
+  if(inputSample_.getSize() != outputSample_.getSize()) throw InvalidDimensionException(HERE) << "Input and output samples must have the same size";
+
+  Description input(1);
+  input[0] = "x";
+  SymbolicFunction one(input, Description(1, "1"));
+  weightFunction_  = one ;
+}
+
+/** Virtual constructor */
+HSICEstimatorImplementation * HSICEstimatorImplementation::clone() const
+{
+  return new HSICEstimatorImplementation(*this);
+}
+
+/** Compute the weight matrix from the weight function */
+SquareMatrix HSICEstimatorImplementation::computeWeightMatrix(const Sample & Y) const
+{
+  const Sample wY = weightFunction_(Y);
+  Scalar meanWY = wY.computeMean()[0];
+  SquareMatrix mat(n_);
+  for(UnsignedInteger j = 0; j < n_; ++j)
+  {
+    mat(j, j) = wY(j, 0) / meanWY;
+  }
+  return mat;
+}
+
+/** Compute a HSIC index (one marginal) by using the underlying estimator (biased or not) */
+Scalar HSICEstimatorImplementation::computeHSICIndex( const Sample & inSample, const Sample & outSample, const CovarianceModel & inCovariance, const CovarianceModel & outCovariance, const SquareMatrix & weightMatrix) const
+{
+  return estimatorType_.computeHSICIndex(inSample, outSample, inCovariance, outCovariance, weightMatrix);
+}
+
+/** Compute HSIC and R2-HSIC indices */
+void HSICEstimatorImplementation::computeIndices()
+{
+  /* Compute weigts */
+  SquareMatrix W = computeWeightMatrix(outputSample_);
+
+  /* Init */
+  HSIC_XX_ = Point(inputDimension_);
+  HSIC_XY_ = Point(inputDimension_);
+  HSIC_YY_ = Point(1);
+
+  /* Loop over marginals: HSIC indices */
+  for(UnsignedInteger dim = 0; dim < inputDimension_; ++dim)
+  {
+    Sample xdim(inputSample_.getMarginal(dim));
+    HSIC_XY_[dim] = computeHSICIndex(xdim, outputSample_, covarianceList_[dim], covarianceList_[inputDimension_], W);
+    HSIC_XX_[dim] = computeHSICIndex(xdim, xdim, covarianceList_[dim], covarianceList_[dim], W);
+  }
+  HSIC_YY_[0] = computeHSICIndex(outputSample_, outputSample_, covarianceList_[inputDimension_], covarianceList_[inputDimension_], W);
+
+  /* Compute R2-HSIC */
+  R2HSICIndices_ = Point(inputDimension_);
+  for(UnsignedInteger dim = 0; dim < inputDimension_; ++dim)
+  {
+    R2HSICIndices_[dim] = HSIC_XY_[dim] / sqrt(HSIC_XX_[dim] * HSIC_YY_[0]);
+  }
+
+}
+
+/** Set permutation size */
+void HSICEstimatorImplementation::setPermutationSize(const UnsignedInteger B)
+{
+  if(B != permutationSize_)
+  {
+    permutationSize_ = B;
+    PValuesPermutation_ = Point();
+  }
+}
```

Comment:
If the `computePValuesPermutation` method is made private, as suggested by @josephmure , the control on whether the user has actually modified the permutation size value should not be done, and the `PValuesPermutation_` value should be reset in any case :
```suggestion
    permutationSize_ = B;
    PValuesPermutation_ = Point();
}
```

The underlying idea here is to allow the user to compute the permutation-based p-values several times with the same parameters in order to study its variability (e.g., when trying to identify an optimal value of `permutationSize_` )

Reply:
Agreed! Sorry Kieran for suggesting this poorly thought-out check. Small precision: `computePValuesPermutation` needs to be "protected", not "private".

Reply:
OK done